### PR TITLE
[ADD] Added web_validation_dialog module.

### PR DESCRIPTION
**Pull request by sudhir-serpentcs**
Added web_security_dialog module.
Renamed it by "**web_validation_dialog**".
Please review.

### CODE REVIEW

**Comment by pedrobaeza** [comment]
You have to remove all the branded images with SerpentCS logo.

**Comment by sudhir-serpentcs** [comment]
>pedrobaeza commented 3 minutes ago 
You have to remove all the branded images with SerpentCS logo.
--

Thanks @pedrobaeza for the feedback. Sure I will remove them.

**Comment by pedrobaeza** [comment]
Check also Travis result.

**Comment by hbrunn** [comment]
you hopefully are aware that this has nothing to do with security but just is a UI restriction? If not, now you are.

The readme should make it very clear that this is the level of "protection" `<field name="name" groups="group" />` provides.

**Comment by sudhir-serpentcs** [comment]
@pedrobaeza Removed branded images and fixed travis.

**Comment by sudhir-serpentcs** [comment]
> hbrunn commented 18 hours ago • edited 
you hopefully are aware that this has nothing to do with security but just is a UI restriction? If not, now you are. The readme should make it very clear that this is the level of "protection" <field name="name" groups="group" /> provides.
--

@hbrunn Should we rename the module name or change the conent in README file? What do you suggest?

**Comment by hbrunn** [comment]
both. Or do it right and implement the pincode for example as a decorator on the button handler that only executes the handler if the correct pincode came along.

**Comment by sudhir-serpentcs** [comment]
@hbrunn @pedrobaeza 
Will that be fine if I rename it by **web_confirmation_dialog** or **web_validation_dialog** as it ask the password to validate the user when user clicks on a button?

**Comment by hbrunn** [comment]
I think that's fine, because then it's clear that this is just meant to prevent human error

**Comment by sudhir-serpentcs** [comment]
@hbrunn Thanks for your quick response. I really appreciate that.

File: web_security_dialog/models/res_company.py
Comment (hbrunn) [inline web_security_dialog/models/res_company.py]: you don't need to search when you already have the id. Use `self.browse`

File: web_security_dialog/static/src/js/web_security_dialog.js
Comment (hbrunn) [inline web_security_dialog/static/src/js/web_security_dialog.js]: better use https://github.com/OCA/OCB/blob/10.0/addons/web/static/src/js/views/form_common.js#L662

File: web_security_dialog/static/src/js/web_security_dialog.js
Comment (hbrunn) [inline web_security_dialog/static/src/js/web_security_dialog.js]: if you use the above, options always will be an object, and then you can lose most of those checks here

**Review by hbrunn — CHANGES_REQUESTED** [review]

File: web_security_dialog/models/res_company.py
Comment (hbrunn) [reply web_security_dialog/models/res_company.py]: ah, and don't forget to write a test for this function to satisfy the codecov CI

File: web_security_dialog/static/src/js/web_security_dialog.js
Comment (sudhir-serpentcs) [reply web_security_dialog/static/src/js/web_security_dialog.js]: WIP

File: web_security_dialog/static/src/js/web_security_dialog.js
Comment (sudhir-serpentcs) [reply web_security_dialog/static/src/js/web_security_dialog.js]: WIP

File: web_security_dialog/models/res_company.py
Comment (sudhir-serpentcs) [reply web_security_dialog/models/res_company.py]: Done

**Comment by sudhir-serpentcs** [comment]
@hbrunn Improved the code as per review. Please review and give feedback.

**Review by hbrunn — APPROVED** [review]

File: web_validation_dialog/LICENSE/LICENSE
Comment (elicoidal) [inline web_validation_dialog/LICENSE/LICENSE]: Please remove this file.

File: web_validation_dialog/__manifest__.py
Comment (elicoidal) [inline web_validation_dialog/__manifest__.py]: We usually use short headers for the copyright and license (with link)

File: web_validation_dialog/models/res_company.py
Comment (elicoidal) [inline web_validation_dialog/models/res_company.py]: a detail but variable name would rather be security_code, no? (nice to have)

**Review by elicoidal — APPROVED** [review]
some details (non-blocking)

File: web_validation_dialog/LICENSE/LICENSE
Comment (sudhir-serpentcs) [reply web_validation_dialog/LICENSE/LICENSE]: Done

File: web_validation_dialog/__manifest__.py
Comment (sudhir-serpentcs) [reply web_validation_dialog/__manifest__.py]: Done. Thanks for the review.

**Comment by sudhir-serpentcs** [comment]
@hbrunn @elicoidal @Yajo @pedrobaeza Thanks for your reviews. Please let me know for any change / correction.

**Comment by elicoidal** [comment]
@sudhir-serpentcs for me LGTM. Would be nice to have another review but otherwise we will merge it in a couple of days

**Comment by sudhir-serpentcs** [comment]
@hbrunn @elicoidal @Yajo @pedrobaeza Can you please check the PR and can be merged it if everything is fine?
CC: @JayVora-SerpentCS

File: web_validation_dialog/__manifest__.py
Comment (yajo) [inline web_validation_dialog/__manifest__.py]: I think this is not an application

File: web_validation_dialog/models/res_company.py
Comment (yajo) [reply web_validation_dialog/models/res_company.py]: Yes, and if you do so, no need to specify the string manually.

File: web_validation_dialog/models/res_company.py
Comment (yajo) [inline web_validation_dialog/models/res_company.py:17]: This method is overly complicated.

1. Do it in an `@api.multi`, where your `self` already is a company, so you don't have to `browse` again.
2. Arguments must be `(self, password)`, not a dict of vals (in fact, **never** do that in Python; we are not in 💩-y old JS, we have keyword arguments for that 😜)
3. Do not transcode password.

The rest of changes should come naturally once you do that.

File: web_validation_dialog/static/src/js/web_validation_dialog.js
Comment (yajo) [inline web_validation_dialog/static/src/js/web_validation_dialog.js]: Please use proper indentation all across this file.

File: web_validation_dialog/static/src/js/web_validation_dialog.js
Comment (yajo) [inline web_validation_dialog/static/src/js/web_validation_dialog.js]: Can't you declare this as a class-level method?

File: web_validation_dialog/static/src/js/web_validation_dialog.js
Comment (yajo) [inline web_validation_dialog/static/src/js/web_validation_dialog.js:128]: This obviously has to change once you change the backend method signature.

**Review by yajo — CHANGES_REQUESTED** [review]

File: web_validation_dialog/__manifest__.py
Comment (sudhir-serpentcs) [reply web_validation_dialog/__manifest__.py]: Done

File: web_validation_dialog/static/src/js/web_validation_dialog.js
Comment (sudhir-serpentcs) [reply web_validation_dialog/static/src/js/web_validation_dialog.js]: Done

File: web_validation_dialog/static/src/js/web_validation_dialog.js
Comment (sudhir-serpentcs) [reply web_validation_dialog/static/src/js/web_validation_dialog.js]: Done

File: web_validation_dialog/models/res_company.py
Comment (sudhir-serpentcs) [reply web_validation_dialog/models/res_company.py]: @Yajo Is there a way to call `api.multi` method from JS?

File: web_validation_dialog/static/src/js/web_validation_dialog.js
Comment (tarteo) [inline web_validation_dialog/static/src/js/web_validation_dialog.js]: Is this return really needed? as well as the other returns in the click cb function?

**Review by tarteo — COMMENTED** [review]
My comments are non-blocking.

File: web_validation_dialog/__manifest__.py
Comment (simahawk) [inline web_validation_dialog/__manifest__.py]: https://github.com/OCA/web

File: web_validation_dialog/models/__init__.py
Comment (simahawk) [inline web_validation_dialog/models/__init__.py]: `s/(c)/Copyright`

File: web_validation_dialog/static/src/js/web_validation_dialog.js
Comment (simahawk) [inline web_validation_dialog/static/src/js/web_validation_dialog.js]: s/From/Form

File: web_validation_dialog/static/src/js/web_validation_dialog.js
Comment (simahawk) [inline web_validation_dialog/static/src/js/web_validation_dialog.js]: bad indent

File: web_validation_dialog/static/src/js/web_validation_dialog.js
Comment (simahawk) [inline web_validation_dialog/static/src/js/web_validation_dialog.js]: I see too many nested calls. IMHO they should be moved to functions of the widget.

File: web_validation_dialog/static/src/js/web_validation_dialog.js
Comment (simahawk) [inline web_validation_dialog/static/src/js/web_validation_dialog.js]: not a callback but a deferred object

File: web_validation_dialog/static/src/xml/web_validation_dialog.xml
Comment (simahawk) [inline web_validation_dialog/static/src/xml/web_validation_dialog.xml:5]: add another meaningful class to ease styling

**Review by simahawk — CHANGES_REQUESTED** [review]

File: web_validation_dialog/static/src/js/web_validation_dialog.js
Comment (simahawk) [reply web_validation_dialog/static/src/js/web_validation_dialog.js]: this one is not blocking but if not addressed I'd like to see a `TODO` comment to remember about it.

**Comment by JayVora-SerpentCS** [comment]
Working soon back on it!

**Comment by JayVora-SerpentCS** [comment]
@simahawk @Yajo

File: web_validation_dialog/README.rst
Comment (yajo) [inline web_validation_dialog/README.rst]: This should be formatted as code:

    ``.....``

**Review by yajo — APPROVED** [review]
Code looks OK.

Please review the JS linter, though.

OTOH, I'm thinking on the "concept" of this addon, and I feel like it's not needed. It seems like you want to forbid access to some buttons to persons who know a secret code for a company, but why don't you simply use permissions? On any button you define you can set the `groups` attribute, which will hide the button for those without that group:

```xml
<button action="whatever" groups="my_addon.group_1,my_addon.group_2"/>
```

:point_up: That way, those that do not have that group will not see the button, and those that have the access will not need to type a password. What does this add on top of that? :thinking:

**Review by simahawk — CHANGES_REQUESTED** [review]
most of the important comments are still unattended :man_shrugging:

File: web_validation_dialog/static/src/js/web_validation_dialog.js
Comment (nikul-serpentcs) [reply web_validation_dialog/static/src/js/web_validation_dialog.js]: @tarteo I think is not needed but if open Dialog alert box then return something(is good for us). so we added return.

File: web_validation_dialog/static/src/xml/web_validation_dialog.xml
Comment (nikul-serpentcs) [reply web_validation_dialog/static/src/xml/web_validation_dialog.xml:5]: @simahawk I used another class but Dialog look ugly, So we keep this.

If you know about another meaningful class then info me.

File: web_validation_dialog/static/src/js/web_validation_dialog.js
Comment (tarteo) [reply web_validation_dialog/static/src/js/web_validation_dialog.js]: You don't return anything. So the way you use `return;` (on this line and line 83 and 92) it does the same as no return statement. Empty returns are mostly used to 'escape' execution of the rest of the function.

File: web_validation_dialog/static/src/js/web_validation_dialog.js
Comment (tarteo) [inline web_validation_dialog/static/src/js/web_validation_dialog.js]: return statement has no meaning here.

File: web_validation_dialog/static/src/js/web_validation_dialog.js
Comment (tarteo) [inline web_validation_dialog/static/src/js/web_validation_dialog.js]: return statement has no meaning here.

File: web_validation_dialog/static/src/js/web_validation_dialog.js
Comment (tarteo) [inline web_validation_dialog/static/src/js/web_validation_dialog.js]: return statement has no meaning here.

**Review by tarteo — CHANGES_REQUESTED** [review]
Can you also review the jslint, there are many errors.

**Comment by nikul-serpentcs** [comment]
@tarteo Done

**Comment by JayVora-SerpentCS** [comment]
@nikul-serpentcs fixup commits when done, good to keep minimum commits.

**Comment by nikul-serpentcs** [comment]
@simahawk @tarteo @pedrobaeza Can you please review code?

File: web_validation_dialog/static/src/xml/web_validation_dialog.xml
Comment (simahawk) [reply web_validation_dialog/static/src/xml/web_validation_dialog.xml:5]: `<div class="validation_dialog container">`?

File: web_validation_dialog/models/res_company.py
Comment (simahawk) [reply web_validation_dialog/models/res_company.py:17]: nothing to do here?

File: web_validation_dialog/static/src/js/web_validation_dialog.js
Comment (simahawk) [inline web_validation_dialog/static/src/js/web_validation_dialog.js:35]: in this case you don't return anything, is this correct given that you expect a callback here `this.view.recursive_save().then(self.exec_action())`?

File: web_validation_dialog/static/src/js/web_validation_dialog.js
Comment (simahawk) [reply web_validation_dialog/static/src/js/web_validation_dialog.js:128]: nothing to do here?

File: web_validation_dialog/static/src/xml/web_validation_dialog.xml
Comment (simahawk) [inline web_validation_dialog/static/src/xml/web_validation_dialog.xml]: this input should have a `name`

File: web_validation_dialog/tests/test_web_validation_dialog.py
Comment (simahawk) [inline web_validation_dialog/tests/test_web_validation_dialog.py:14]: no test for the invalid case?

**Review by simahawk — CHANGES_REQUESTED** [review]

File: web_validation_dialog/models/res_company.py
Comment (nikul-serpentcs) [reply web_validation_dialog/models/res_company.py]: @simahawk Here only check Password right or not.

This method call by JS [here](https://github.com/OCA/web/pull/728/files/ef07f9af45331733aba9bfb517991c0da7591a5e#diff-820648e75e6f8a70cf37174d21a94c2bR108)

File: web_validation_dialog/static/src/js/web_validation_dialog.js
Comment (nikul-serpentcs) [reply web_validation_dialog/static/src/js/web_validation_dialog.js:35]: @simahawk here If Password is confirmed then Open change Password view, No need to return anything.

peek 2018-11-16 12-09

File: web_validation_dialog/models/res_company.py
Comment (nikul-serpentcs) [reply web_validation_dialog/models/res_company.py]: @simahawk Here only check Password right or not.

This method call by JS [here](https://github.com/OCA/web/pull/728/files/ef07f9af45331733aba9bfb517991c0da7591a5e#diff-820648e75e6f8a70cf37174d21a94c2bR108)

Comment (nikul-serpentcs) [reply web_validation_dialog/models/res_company.py:17]: > nothing to do here?

@simahawk Here only check Password right or not.

This method call by JS [here](https://github.com/OCA/web/pull/728/files/ef07f9af45331733aba9bfb517991c0da7591a5e#diff-820648e75e6f8a70cf37174d21a94c2bR108)

File: web_validation_dialog/static/src/js/web_validation_dialog.js
Comment (nikul-serpentcs) [reply web_validation_dialog/static/src/js/web_validation_dialog.js:128]: > nothing to do here?

@simahawk This method Call in python method [here](https://github.com/OCA/web/pull/728/files/ef07f9af45331733aba9bfb517991c0da7591a5e#diff-6113c0a76b7ebbb8aca53ab6c3696d75R13)

It's Check Only Password and return here.

File: web_validation_dialog/tests/test_web_validation_dialog.py
Comment (nikul-serpentcs) [reply web_validation_dialog/tests/test_web_validation_dialog.py:14]: @simahawk Added invalid case.

**Comment by nikul-serpentcs** [comment]
@simahawk Improved Code as per your comment, and recomment your question, Could you please check and move forward.

File: web_validation_dialog/static/src/js/web_validation_dialog.js
Comment (daramousk) [inline web_validation_dialog/static/src/js/web_validation_dialog.js:87]: There is no need to manually `blockUI()` or `unblockUI()`, this happens automatically if your rpc does not return within 3s. Implemented on `web.Loading`

File: web_validation_dialog/static/src/js/web_validation_dialog.js
Comment (daramousk) [inline web_validation_dialog/static/src/js/web_validation_dialog.js:86]: You can mark a field as `required="required"` on the html instead of writing code to handle it.

File: web_validation_dialog/static/src/js/web_validation_dialog.js
Comment (daramousk) [inline web_validation_dialog/static/src/js/web_validation_dialog.js:84]: That does not seem to be necessary.

File: web_validation_dialog/static/src/js/web_validation_dialog.js
Comment (daramousk) [reply web_validation_dialog/static/src/js/web_validation_dialog.js:128]: I think it would be more appropriate to actually add `.done()`, `.fail()` handlers here instead of defining them every time after the usage of this function.

File: web_validation_dialog/static/src/js/web_validation_dialog.js
Comment (daramousk) [inline web_validation_dialog/static/src/js/web_validation_dialog.js:91]: Same here, maybe these `fail`, `done` handlers should be attached to the `validate_pincode` function itself so you can remove those `if/else Dialog.alert`

**Review by daramousk — CHANGES_REQUESTED** [review]
Hello there,

I posted some comments on this and I will like to add some general remarks (most have already been mentioned).

1) The people who are going to use this will expect that people who do not have the passcode for a specific rpc call/button click, will not be able to invoke that call. That is not true here since anyone can do anything to the js side of things.

2) With that being said, maybe something analogous could be done using exclusively python. 
You could very well, return a dialog to get input (action with window=new` or something else, validate that input on the backend and return an error message (return {'warning' ...) or proceed calling the intended action.

**Pull request closed without merging**
